cpp/src/arrow/array/array_test.cc

```diff
@@ -1307,6 +1307,14 @@ TEST(TestBooleanArray, TrueCountFalseCount) {
   CheckArray(checked_cast<const BooleanArray&>(*arr));
   CheckArray(checked_cast<const BooleanArray&>(*arr->Slice(5)));
   CheckArray(checked_cast<const BooleanArray&>(*arr->Slice(0, 0)));
+
+  // GH-41016 true_count() with array without validity buffer with null_count of -1
+  auto data = ArrayFromJSON(boolean(), "[true, false, true]")->data();
+  data->null_count = -1;
+  std::shared_ptr<BooleanArray> arr2(new BooleanArray(data));
```

Comment:
Can we use `make_shared` here?

```suggestion
  auto arr2 = std::make_shared<BooleanArray>(data);
```

Can we use more meaningful variable name than `arr2` like `arr_unknown_null_count` or something?

Reply:
Thanks, that's cleaner!